File: Problems 0100-0199/112. Path Sum/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root:
            return False
        return self.solve(root, 0, targetSum)

    def solve(self, root, currSum, targetSum):
        if not root:
            return False

        if not root.left and not root.right:
            return currSum + root.val == targetSum

        return self.solve(root.left, currSum + root.val, targetSum) or \
        self.solve(root.right, currSum + root.val, targetSum)

File: Problems 0100-0199/112. Path Sum/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 class Solution {
    public:
        bool hasPathSum(TreeNode* root, int targetSum) {
            return hasPath(root, 0, targetSum);
        }
    
        bool hasPath(TreeNode* root, int currSum, int& targetSum) {
            if (!root) return false;
    
            currSum += root->val;
            if (!root->left && !root->right) {
                return currSum == targetSum;
            }
    
            return hasPath(root->left, currSum, targetSum) || hasPath(root->right, currSum, targetSum);
        }
    };